autorecsys/pipeline/preprocessor.py

```diff
@@ -78,6 +103,216 @@ def preprocessing(self, **kwargs):
+        categ_count_dict = defaultdict(lambda: defaultdict(int))
+
+        with open(self.dataset_path, 'r') as ld:
+            for line in ld:
+                # Using zero "0" as substitute for missing values in:
+                #   1) Numerical feature may corrupt data because 0 is numerical.
+                #   2) Categorical feature will not corrupt data because the string "0" is a new category.
+                # TODO: Need a principled way to determine missing values for numerical features.
+                line = [v if v != "" else "0" for v in line.strip('\n').split('\t')]
+
+                for i, j in enumerate(self.label_indices):  # record label feature data
+                    label_ll[i].append(int(line[j]))  # typecast from string saves memory
+                for i, j in enumerate(self.numer_indices):  # record numerical feature data
+                    numer_ll[i].append(int(line[j]))  # typecast from string saves memory
+                for i, j in enumerate(self.categ_indices):  # record categorical feature data
+                    categ_ll[i].append(line[j])  # cannot typecast string data
+                    categ_count_dict[i][line[j]] += 1  # count category occurrences
+
+        # Step 2: Create dictionary for indexing categories.
+        if Path(self.fit_dictionary_path).is_file():
+            categ_index_dict = load_pickle(self.fit_dictionary_path)
+        else:
+            # Define defaultdict where key="feature index" & val="defaultdict" where key="category" & val="index".
+            categ_index_dict = defaultdict(lambda: defaultdict(int))
+
+            for feat_index, categ_dict in categ_count_dict.items():
+                categ_index = 0
+                for categ, count in categ_dict.items():
+                    if count >= self.categ_filter:  # index filtered categories at a later stage
+                        categ_index_dict[feat_index][categ] = categ_index
+                        categ_index += 1
+            del categ_count_dict  # release memory
+            gc.collect()
+            # Save fit dictionary.
+            save_pickle(self.fit_dictionary_path, dict(categ_index_dict))  # cannot pickle defaultdict using lambda
+
+        # Step 3: Index categories.
+        for i, categ_l in enumerate(categ_ll):
+            for j, c in enumerate(categ_l):
+                if c in categ_index_dict[i]:
+                    categ_ll[i][j] = categ_index_dict[i][c]  # index in-place
+                else:  # index filtered categories as the end index
+                    categ_ll[i][j] = len(categ_index_dict[i])
+        del categ_index_dict  # release memory
+        gc.collect()
+
+        # Step 4: Format data and obtain hash statistics.
+        array_data = np.concatenate((np.asarray(label_ll).T, np.asarray(numer_ll).T, np.asarray(categ_ll).T), axis=1)
+        del label_ll  # release memory
```

Comment:
I think you don't have to call del multiple times, just do one time in the end like ``del a,b,c``

Reply:
Agreed, we can combine consecutive / line-after-line memory free statements into the same line.